cli/internal/daemon/connector/connector.go

```diff
@@ -0,0 +1,320 @@
+	// hopefully never happen.
+	lockFile, err := lockfile.New(c.PidPath.ToString())
+	if err != nil {
+		return err
+	}
+	process, err := lockFile.GetOwner()
+	if err == nil {
+		// Check that this is the same process that we failed to connect to.
+		// Otherwise, loop around again and start with whatever new process has
+		// the pid file.
+		if process.Pid == pid {
+			// we have a process that we need to kill
+			// TODO(gsoltis): graceful kill? the process is already not responding to requests,
+			// but it could be in the middle of a graceful shutdown. Probably should let it clean
+			// itself up, and report an error and defer to a force-kill by the user
+			if err := process.Kill(); err != nil {
+				return err
+			}
+		}
+		return nil
+	} else if errors.Is(err, os.ErrNotExist) {
+		// There's no pid file. Someone else killed it. Loop around and try again
+		return nil
+	} else if errors.Is(err, lockfile.ErrDeadOwner) {
+		// The daemon crashed. report an error to the user
+		return err
+	}
+	return err
+}
+
+// Connect attempts to create a connection to a turbo daemon.
+// Retries and daemon restarts are built in. If this fails,
+// it is unlikely to succeed after an automated retry.
+func (c *Connector) Connect(ctx context.Context) (*Client, error) {
+	client, err := c.connectInternal(ctx)
+	if err != nil {
+		return nil, c.wrapConnectionError(err)
+	}
+	return client, nil
+}
+
+func (c *Connector) connectInternal(ctx context.Context) (*Client, error) {
+	for i := 0; i < _maxAttempts; i++ {
+		lockFile, err := lockfile.New(c.PidPath.ToString())
+		if err != nil {
+			// Should only happen if we didn't pass an absolute path
+			return nil, err
+		}
+		var serverPid int
+		if daemonProcess, err := lockFile.GetOwner(); errors.Is(err, lockfile.ErrDeadOwner) {
+			// Report an error? We could technically race with another client trying to
+			// start a daemon here.
+			return nil, errors.Wrap(err, "pid file appears stale. If no daemon is running, please remove it")
+		} else if os.IsNotExist(err) {
+			if c.Opts.DontStart {
+				return nil, ErrDaemonNotRunning
+			}
+			// The pid file doesn't exist. Start a daemon
+			pid, err := c.startDaemon()
+			if err != nil {
+				return nil, err
+			}
+			serverPid = pid
+		} else {
+			serverPid = daemonProcess.Pid
+		}
+		if err := c.waitForSocket(); errors.Is(err, ErrFailedToStart) {
+			continue
+		} else if err != nil {
+			return nil, err
+		}
+		client, err := c.getClientConn()
+		if err != nil {
+			return nil, err
+		}
+		if err := c.sendHello(ctx, client); err == nil {
+			// We connected and negotiated a version, we're all set
+			return client, nil
+		} else if errors.Is(err, errVersionMismatch) {
+			// killLiveServer is responsible for closing the client
+			if err := c.killLiveServer(ctx, client, serverPid); err != nil {
+				return nil, err
+			}
+		} else if errors.Is(err, errConnectionFailure) {
+			// close the client, see if we can kill the stale daemon
```

Comment:
`killLiveServer` being responsible for killing the client and `killDeadServer` not being responsible for it is a bit of incongruence. I'm glad it's commented, but can we get rid of that incongruence? Or make it seem like they'll be less congruent by changing the names of the methods?

Reply:
Hmm. `killLiveServer` is _also_ responsible for closing the client, since it needs to use it. With a little bit of juggling we could change that, I think. Something like:

```go
err := c.killLiveServer(ctx, client, serverPid)
_ = client.Close()
if err != nil {
  ...
}
```

Or do you think we're better off with some different naming or comments?

Reply:
I updated the comment around this to hopefully be more clear.

Reply:
I don't have strong opinions here, I just was noting the incongruence. As long as it isn't a Home Alone-like trap in the codebase I'm okay.

Reply:
Ok, marking this resolved. Happy to update comment if needed